weavertest/deployer.go

```diff
@@ -192,68 +180,69 @@ func (d *deployer) Init(config string) weaver.Instance {
 		SingleMachine: d.wlet.SingleMachine,
 		RunMain:       true,
 	}
-	// TODO(mwhittaker): Issue an UpdateComponents call to the main group.
-	// Right now, the main co-location group knows to start main, but in the
-	// future, it won't.
+	var e *conn.EnvelopeConn
+	created := make(chan struct{})
 	go func() {
 		// NOTE: We must create the envelope conn in a separate gorotuine
 		// because it initiates a blocking handshake with the weavelet
 		// (initialized below via weaver.Init).
-		g := d.group("main")
-		handler := &handler{
-			deployer:   d,
-			group:      g,
-			subscribed: map[string]bool{},
-		}
-		conn, err := conn.NewEnvelopeConn(fromWeaveletReader, toWeaveletWriter, handler, wlet)
-		if err != nil {
+		var err error
+		if e, err = conn.NewEnvelopeConn(
+			d.ctx, fromWeaveletReader, toWeaveletWriter, wlet); err != nil {
 			panic(fmt.Errorf("cannot create envelope conn: %w", err))
 		}
-		handler.conn = connection{conn: conn}
-		if err := d.registerReplica(g, conn.WeaveletInfo()); err != nil {
-			panic(fmt.Errorf(`cannot register the replica for "main"`))
-		}
-
-		// TODO(mwhittaker): Close this conn when the unit test ends. Right
-		// now, the conn lives forever. This means the pipes are also leaking.
-		// We might have to add a Close method to EnvelopeConn.
-		if err := conn.Serve(); err != nil {
-			d.t.Error(err)
-		}
+		created <- struct{}{}
 	}()
-
 	bootstrap := runtime.Bootstrap{
 		ToWeaveletFile: toWeaveletReader,
 		ToEnvelopeFile: fromWeaveletWriter,
 		TestConfig:     config,
 	}
 	ctx := context.WithValue(d.ctx, runtime.BootstrapKey{}, bootstrap)
-	return weaver.Init(ctx)
+	instance := weaver.Init(ctx)
+	<-created
+
+	g := d.group("main")
+	g.conns = append(g.conns, connection{conn: e})
+	handler := &handler{
+		deployer:   d,
+		group:      g,
+		subscribed: map[string]bool{},
+		conn:       connection{conn: e},
+	}
+	e.Serve(handler)
+	d.running.Go(func() error {
+		err := e.Wait()
+		d.stop(err)
+		return err
+	})
+
+	if err := d.registerReplica(g, e.WeaveletInfo()); err != nil {
+		panic(fmt.Errorf(`cannot register the replica for "main"`))
+	}
+	// TODO(mwhittaker): Issue an UpdateComponents call to the main group.
```

Comment:
You can get rid of this todo. I added a `RunMain` field to the `WeaveletSetupInfo`.